src/Microsoft.AspNet.Security.OpenIDConnect/OpenIdConnectAuthenticationMiddleware.cs

```diff
@@ -0,0 +1,173 @@
+using Microsoft.AspNet.Security.DataHandler.Serializer;
+using Microsoft.AspNet.Security.DataProtection;
+using Microsoft.AspNet.Security.Infrastructure;
+using Microsoft.Framework.Logging;
+using Microsoft.Framework.OptionsModel;
+using Microsoft.IdentityModel.Protocols;
+
+namespace Microsoft.AspNet.Security.OpenIdConnect
+{
+    /// <summary>
+    /// ASP.NET middleware for obtaining identities using OpenIdConnect protocol.
+    /// </summary>
+    public class OpenIdConnectAuthenticationMiddleware : AuthenticationMiddleware<OpenIdConnectAuthenticationOptions>
+    {
+        private readonly ILogger _logger;
+
+        /// <summary>
+        /// Initializes a <see cref="OpenIdConnectAuthenticationMiddleware"/>
+        /// </summary>
+        /// <param name="next">The next middleware in the ASP.NET pipeline to invoke</param>
+        /// <param name="app">The ASP.NET application</param>
+        /// <param name="options">Configuration options for the middleware</param>
+        [SuppressMessage("Microsoft.Reliability", "CA2000:Dispose objects before losing scope", Justification = "Managed by caller")]
+        public OpenIdConnectAuthenticationMiddleware(
+            RequestDelegate next,
+            IServiceProvider services,
+            IDataProtectionProvider dataProtectionProvider,
+            ILoggerFactory loggerFactory,
+            IOptions<ExternalAuthenticationOptions> externalOptions,
+            IOptions<OpenIdConnectAuthenticationOptions> options,
+            ConfigureOptions<OpenIdConnectAuthenticationOptions> configureOptions)
+            : base(next, services, options, configureOptions)
+        {
+            _logger = loggerFactory.Create<OpenIdConnectAuthenticationMiddleware>();
+
+            if (string.IsNullOrWhiteSpace(Options.TokenValidationParameters.AuthenticationType))
+            {
+                Options.TokenValidationParameters.AuthenticationType = externalOptions.Options.SignInAsAuthenticationType;
+            }
+
+            if (Options.StateDataFormat == null)
+            {
+                var dataProtector = dataProtectionProvider.CreateDataProtector(
+                    typeof(OpenIdConnectAuthenticationMiddleware).FullName, 
+					typeof(string).FullName,
+                    Options.AuthenticationType,
+					"v1");
+
+                Options.StateDataFormat = new PropertiesDataFormat(dataProtector);
+            }
+
+            if (Options.StringDataFormat == null)
+            {
+                var dataProtector = dataProtectionProvider.CreateDataProtector(
+                    typeof(OpenIdConnectAuthenticationMiddleware).FullName,
+                    typeof(string).FullName,
+                    Options.AuthenticationType,
+                    "v1");
+
+                Options.StringDataFormat = new SecureDataFormat<string>(new StringSerializer(), dataProtector, TextEncodings.Base64Url);
+            }
+
+            if (Options.SecurityTokenValidators == null)
```

Comment:
agree as above, planning on making the change to return an empty collection. Just ran out of time for this set of changes.